src/site/content/en/blog/airshift-perf-optimization/index.md

```diff
@@ -0,0 +1,250 @@
+ <figcaption class="w-figcaption w-figcaption--fullbleed">
+    Endless spinner on low end devices
+  </figcaption>
+</figure>
+
+On the other hand, AifSHIFT had many expensive scripts that ran on the device, which was consuming the majority of the main thread activity, but the engineering team did not realize how harmful it was because they were developing and testing on a rich spec computer with fast WiFi.
+
+<figure class="w-figure w-figure--fullbleed">
+  <img class="w-screenshot" src="main-thread-break-down.png" alt="A chart that shows when loading the shift table, around 80% of the load time was consumed by running scripts">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    A chart that shows when loading the shift table, around 80% of the load time was consumed by running scripts
+  </figcaption>
+</figure>
+
+It was now obvious that performance optimization was needed. AirSHIFT formed a task force to tackle this issue and here are 5 things they focused on to make their app more responsive.
+
+
+## **1. Virtualize large tables**
+Displaying the shift table required multiple expensive steps: constructing the virtual DOM and rendering it on screen in proportion to the number of staff members and time slots. For example, if a restaurant had 50 working members and wanted to check their monthly shift schedule, it would be a table of 50 (members) multiplied by 30 (days) which would lead to 1,500 cell components to render. This is a very expensive operation especially for low spec devices. In reality, things get worse. From the research they understood there were shops that need to manage around 200 staff members requiring around 6,000 cell components in the single monthly table.
+
+To reduce the cost of this operation, AirSHIFT virtualized the shift table. This means that it will only mount the components within the viewport and unmount any of the off screen components. 
+
+<figure class="w-figure w-figure--fullbleed">
+  <img class="w-screenshot" src="virtualize_before.png" alt="Before: rendering all the shift table cells">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    Before: rendering all the shift table cells
+  </figcaption>
+</figure>
+
+<figure class="w-figure w-figure--fullbleed">
+  <img class="w-screenshot" src="virtualize_after.png" alt="After: only rendering the cells within the viewport">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    After: only rendering the cells within the viewport
+  </figcaption>
+</figure>
+
+In this case, AirSHIFT used [react-virtualized](https://github.com/bvaughn/react-virtualized) as there were requirements around enabling complex two dimensional grid table. They are also exploring ways to convert the implementation to use the lightweight [react-window](https://github.com/bvaughn/react-window) in the future.
+
+{% Aside %}
+Related article: [Virtualize large lists with react-window](https://web.dev/virtualize-long-lists-react-window/)
+{% endAside %}
```

Comment:
Please remove this aside now that we're already linking to this article

Reply:
removed.